subsys/net/lib/coap/coap_client.c

```diff
@@ -521,6 +533,11 @@ static void coap_client_resend_handler(struct coap_client *client)
 
 	for (int i = 0; i < CONFIG_COAP_CLIENT_MAX_REQUESTS; i++) {
 		if (timeout_expired(&client->requests[i])) {
+			if (!client->requests[i].coap_request.confirmable) {
+				release_internal_request(&client->requests[i]);
```

Comment:
Perhaps we can still belly up to application that this request has been internally released due to no-response, so the appropriate cleanup can happen.

`report_callback_error(&client->requests[i], -ENOMSG);`